libraries/src/Event/Result/ResultTypeObjectAware.php

```diff
@@ -0,0 +1,111 @@
+<?php
+/**
+ * Joomla! Content Management System
+ *
+ * @copyright  (C) 2022 Open Source Matters, Inc. <https://www.joomla.org>
+ * @license        GNU General Public License version 2 or later; see LICENSE
+ */
+
+namespace Joomla\CMS\Event\Result;
+
+\defined('JPATH_PLATFORM') or die;
+
+use InvalidArgumentException;
+
+/**
+ * This Trait partially implements the ResultAwareInterface for type checking.
+ *
+ * Events using this Trait (and the ResultAware trait) will expect event handlers to set results
+ * of an object type.
+ *
+ * If you do not set a list of acceptable result classes any PHP object will satisfy this type check.
+ *
+ * @since  __DEPLOY_VERSION__
+ */
+trait ResultTypeObjectAware
+{
+	/**
+	 * Can the result attribute values also be NULL?
+	 *
+	 * @var    boolean
+	 * @since  __DEPLOY_VERSION__
+	 */
+	protected $resultIsNullable = false;
+
+	/**
+	 * Can the result attribute values also be boolean FALSE?
+	 *
+	 * @var    boolean
+	 * @since  __DEPLOY_VERSION__
+	 *
+	 * @deprecated 5.0 You should use nullable values or exceptions instead of returning boolean false results.
+	 */
+	protected $resultIsFalseable = false;
+
+	/**
+	 * Acceptable class names for result values.
+	 *
+	 * @var    array
+	 * @since  __DEPLOY_VERSION__
+	 */
+	protected $resultAcceptableClasses = [];
+
+	/**
+	 * Checks the type of the data being appended to the result argument.
+	 *
+	 * @param   mixed  $data  The data to type check
+	 *
+	 * @return  void
+	 * @throws  InvalidArgumentException
+	 *
+	 * @internal
+	 * @since   __DEPLOY_VERSION__
+	 */
+	public function typeCheckResult($data): void
+	{
+		if ($this->resultIsNullable && $data === null)
+		{
+			return;
+		}
+
+		if ($this->resultIsFalseable && $data === false)
+		{
+			return;
+		}
+
+		if (empty($this->resultAcceptableClasses))
+		{
+			if (!is_object($data))
+			{
+				throw new InvalidArgumentException(sprintf('Event %s only accepts object results.', $this->getName()));
+			}
+
+			return;
+		}
+
+		$isAcceptable = function ($object)
+		{
+			if (!is_object($object))
+			{
+				return false;
+			}
+
+			foreach ($this->resultAcceptableClasses as $className)
+			{
+				if (is_a($object, $className))
+				{
+					return true;
+				}
+			}
+
+			return false;
+		};
+
+		if (!$isAcceptable($data))
+		{
+			$acceptableTypes = implode(', ', $this->resultAcceptableClasses);
+			$messageTemplate = 'Event %s only accepts object results which are instances of one of %s.';
+			throw new InvalidArgumentException(sprintf($messageTemplate, $this->getName(), $acceptableTypes));
+		}
```

Comment:
```suggestion
		if (!is_object($data))
		{
			throw new InvalidArgumentException(sprintf('Event %s only accepts object results.', $this->getName()));
		}

		if (empty($this->resultAcceptableClasses))
		{
			return;
		}

		foreach ($this->resultAcceptableClasses as $className)
		{
			if (is_a($object, $className))
			{
				return true;
			}
		}

		$acceptableTypes = implode(', ', $this->resultAcceptableClasses);
		$messageTemplate = 'Event %s only accepts object results which are instances of one of %s.';
		throw new InvalidArgumentException(sprintf($messageTemplate, $this->getName(), $acceptableTypes));
```

Keeps it a bit shorter and returns earlier.